From fa534c4a66f5fab7a9c5f9de76c81625e0b64068 Mon Sep 17 00:00:00 2001
From: Yousong Zhou <yszhou4tech@gmail.com>
Date: Thu, 23 Jul 2015 20:47:29 +0800
Subject: [PATCH 9/9] pam_exec: fix build when strndupa() is not available.

 * /modules/pam_exec/pam_exec.c: use strncpy() and local array instead
   of strndupa()

Signed-off-by: Yousong Zhou <yszhou4tech@gmail.com>
---
 modules/pam_exec/pam_exec.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/modules/pam_exec/pam_exec.c b/modules/pam_exec/pam_exec.c
index 17ba6ca..d5ccfeb 100644
--- a/modules/pam_exec/pam_exec.c
+++ b/modules/pam_exec/pam_exec.c
@@ -102,6 +102,7 @@ call_exec (const char *pam_type, pam_handle_t *pamh,
   int use_stdout = 0;
   int optargc;
   const char *logfile = NULL;
+  char *_authtok[PAM_MAX_RESP_SIZE];
   const char *authtok = NULL;
   pid_t pid;
   int fds[2];
@@ -178,11 +179,15 @@ call_exec (const char *pam_type, pam_handle_t *pamh,
 		}
 
 	      pam_set_item (pamh, PAM_AUTHTOK, resp);
-	      authtok = strndupa (resp, PAM_MAX_RESP_SIZE);
+	      _authtok[PAM_MAX_RESP_SIZE-1] = '\0';
+	      authtok = strncpy(_authtok, resp, PAM_MAX_RESP_SIZE-1);
 	      _pam_drop (resp);
 	    }
 	  else
-	    authtok = strndupa (void_pass, PAM_MAX_RESP_SIZE);
+	    {
+	      _authtok[PAM_MAX_RESP_SIZE-1] = '\0';
+	      authtok = strncpy(_authtok, void_pass, PAM_MAX_RESP_SIZE-1);
+	    }
 
 	  if (pipe(fds) != 0)
 	    {
-- 
1.7.10.4

